background/providers/provider.js

```diff
@@ -0,0 +1,186 @@
+'use strict';
+var nconf = require('nconf');
+
+var mongoose = require('mongoose');
+var Category = require('models/Category.js');
+var Torrent = require('models/Torrent.js');
+
+var log = require(__dirname + '/../logging.js');
+var torrentsToAdd = [];
+class Provider {
+    constructor(provider) {
+        // Set the duration a provider should run
+        switch(nconf.get('providers:' + provider + ':config:duration')) {
+            case '@hourly':
+                this.duration = 3600000;
+                break;
+            case '@daily':
+                this.duration = 86400000;
+                break;
+            default:
+                this.duration = Number(nconf.get('providers:' + provider + ':config:duration'));
+        }
+
+        // Set whether or not a provider should run at startup
+        if(typeof(nconf.get('providers:' + provider + ':config:startup')) === 'boolean') {
+            this.runAtStartup = nconf.get('providers:' + provider + ':config:startup');
+        } else {
+            this.runAtStartup = false;
+        }
+    }
+
+    // Checks if a torrent should be added to the database
+    static shouldAddTorrent(category) {
+        if(nconf.get('torrents:whitelist:enabled')) {
+            if(nconf.get('torrents:whitelist:categories').forEach(function(el) {
+                    if(el.toUpperCase() === category.toUpperCase()) {
+                        return true;
+                    }
+             })) {
+                return true;
+            }
+        } else if(nconf.get('torrents:blacklist:enabled')) {
+            if(!nconf.get('torrents:blacklist:categories').forEach(function(el) {
+                    if(el.toUpperCase() === category.toUpperCase()) {
+                        return true;
+                    }
+             })) {
+                return true;
+            }
+        } else {
+            return true;
+        }
+
+        return false;
+    }
+
+    // Add a torrent to the database
+    static addTorrent(title, aliases, size, details, swarm, lastmod, imported, infoHash) {
+        // Validate Data
+        if(typeof(title) !== 'string' || typeof(infoHash) !== 'string') {
```

Comment:
The same goes for here since if it's a private tracker it may not have a infoHash/magnet link from the API.